### PR TITLE
Handle DataTable data reset with InfiniteScroll

### PR DESCRIPTION
**Pull request by leossantiago**
#### What does this PR do?

Provides a fix for #4934, allowing `onMore` to be triggered when manually resetting `DataTable` data.
Even though it's related to `DataTable`, the fix was made on `InfiniteScroll`, as it's the one handling `onMore` calls.

This issue has a use case that updates data dynamically, but when a manual reset is made, some internal states are kept. This fix verifies a condition that indicates a manual data reset (data length = 0, last page = 0, and pending length != 0). If the condition is valid, it resets pageHeights, pendingLength and renderPageBounds, which will make the DataTable behave just as how it works when it first renders.

Note that the given example starts with `currentData = [ ]`, and this should be consistently followed in order for this to work, meaning that the reset should set `currentData = [ ]`.

#### Where should the reviewer start?

`src/js/components/InfiniteScroll/InfiniteScroll.js`

#### What testing has been done on this PR?

Manual testing using storybook, as this issue involves updates to `BoundingCliendRect`, and I do believe those can't be easily tested using RTL/JSDom.

#### How should this be manually tested?

An updated version of the example provided by the issue reporter. Based on storybook/DataTable/Infinite Scroll

https://gist.github.com/leossantiago/cfcbd3de1e88d28c2d96df98b199c4b9

#### What are the relevant issues?

#4934 

#### Screenshots (if appropriate)

Peek 2021-02-08 02-41

#### Is this change backwards compatible or is it a breaking change?

Backwards compatible.

### CODE REVIEW

File: src/js/components/InfiniteScroll/InfiniteScroll.js [hunk 1]
```diff
@@ -147,6 +147,14 @@ const InfiniteScroll = ({
     }
   }, [items.length, lastPage, onMore, pendingLength, renderPageBounds, step]);
 
+  useEffect(() => {
+    if (items.length === 0 && lastPage === 0 && pendingLength !== 0) {
```
Comment (ericsoderberghp) [inline src/js/components/InfiniteScroll/InfiniteScroll.js:151]: Can we just key off of `items.length` here?

File: src/js/components/InfiniteScroll/InfiniteScroll.js [hunk 2]
```diff
@@ -147,6 +147,14 @@ const InfiniteScroll = ({
...
+      setPageHeights([]);
+      setPendingLength(0);
+      setRenderPageBounds([0, 0]);
```
Comment (ericsoderberghp) [inline src/js/components/InfiniteScroll/InfiniteScroll.js:154]: I think we should reset `renderPageBounds` consistent with how we initialize it on line 37.

**Review by ericsoderberghp — COMMENTED** [review]
I think this is a reasonable step to take.

However, at some point we might want to look a little deeper and detect when the items have shrunk such that we don't have enough items for the `pageHeights` we've got. That can be a later change though.

**Comment by leossantiago** [comment]
@ericsoderberghp could you check this once again?

Also, should we keep this open while we investigate `detect when the items have shrunk such that we don't have enough items for the pageHeights we've got` ?

I can keep working on that. Just let me know if we'll be keeping it here, or separating in another issue 😄

File: src/js/components/InfiniteScroll/InfiniteScroll.js [hunk 3]
```diff
@@ -147,6 +147,14 @@ const InfiniteScroll = ({
     }
   }, [items.length, lastPage, onMore, pendingLength, renderPageBounds, step]);
 
+  useEffect(() => {
+    if (lastPage === 0 && pendingLength !== 0) {
+      setPageHeights([]);
+      setPendingLength(0);
+      setRenderPageBounds([0, show ? Math.floor((show + step) / step) - 1 : 0]);
```
Comment (ericsoderberghp) [inline src/js/components/InfiniteScroll/InfiniteScroll.js:154]: I think this is fine functionally. One small improvement would be to create something like:
`const calculateLastPage = (show, step) => (show ? Math.floor((show + step) / step) - 1 : 0);`
outside of the component in this file and then call it in the three places we have that expression.

Comment (leossantiago) [reply src/js/components/InfiniteScroll/InfiniteScroll.js]: Done!
Went with `calculateLastPageBound` as we already have a `lastPage` variable that isn't directly related to this function.

**Review by ericsoderberghp — APPROVED** [review]

**Pull request merged**